src/cellrank/estimators/terminal_states/_gpcca.py

```diff
@@ -532,6 +534,150 @@ def set_initial_states(
         )
         return self
 
+    def get_tsi(self, n_macrostates: int, terminal_states: List[str], cluster_key: str) -> pd.DataFrame:
+        """Compute terminal state identificiation (TSI).
+
+        Parameters
+        ----------
+        n_macrostates
+            Maximum number of macrostates to consider.
+        terminal_states
+            List of terminal states.
+        cluster_key
+            Key in :attr:`~anndata.AnnData.obs` defining cluster labels including terminal states.
+
+        Returns
+        -------
+        Returns TSI as a Pandas DataFrame and adds the class attribute :attr:`tsi`.
+        """
+        macrostates = {}
+        for n_states in range(1, n_macrostates):
+            self.compute_macrostates(n_states=n_states, cluster_key=cluster_key)
+            macrostates[n_states] = self._macrostates.assignment.cat.categories
+
+        max_terminal_states = len(terminal_states)
+
+        tsi_df = {
+            "Number of macrostates": [],
+            "Identified terminal states": [],
+            "Optimal identification": [],
+        }
+        for n_states, states in macrostates.items():
+            n_terminal_states = (
+                states.str.replace(r"(_).*", "", regex=True).drop_duplicates().isin(terminal_states).sum()
+            )
+            tsi_df["Number of macrostates"].append(n_states)
```

Comment:
I slightly prefer to not have spaces in the dataframe's columns and have all lower case, but no strong preference, so we can keep if you want.

Reply:
Done.

Reply:
I understand, but it's not a big change, please consider using again `number_of_macrostates`, etc.); can be done easily with [DataFrame.rename](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rename.html).